src/util.cc

```diff
@@ -630,3 +632,46 @@ bool Truncate(const string& path, size_t size, string* err) {
   }
   return true;
 }
+
+bool ReplaceContent(const string& file_dst, const string& new_content,
+                    string* err) {
+#ifdef _WIN32
+  if (!DeleteFile(file_dst.c_str())) {
+    *err = GetLastErrorString();
+    return false;
+  }
+
+  if (rename(new_content.c_str(), file_dst.c_str()) < 0) {
+    *err = strerror(errno);
+    return false;
+  }
+#else
+  struct stat old_file;
+  bool found_uid_gid = true;
+
+  if (stat(file_dst.c_str(), &old_file) < 0) {
+    found_uid_gid = false;
+  }
+
+  if (unlink(file_dst.c_str()) < 0) {
```

Comment:
This provides a window of opportunity where the file named file_dst is not on the filesystem.

Better would be to take advantage of the posix filesystem guarentee that renaming a file into the name of an already existing file, so long as the destination file can be modified and/or deleted by the current user, is an atomic operation.

So you would want to first create the new file, and give it it's contents. Then change it's attributes, and then rename.

e.g.

```
stat
write_file
change new file attributes
rename
// no unlink
if rename fails, delete new-file before erroring out, to avoid litter.
```